### ISSUE TITLE
QUESTION: How does serverless deal with the removal or rename of actions/feeds/triggers from its config?

### ISSUE THREAD

**Issue by allen-servedio** [serverless/serverless-openwhisk#99]
In looking at use cases in https://github.com/apache/incubator-openwhisk-wskdeploy/blob/master/specification/openwhisk_v0.9.1.pdf (around line number 674), the spec talks about project sync concerns. Is there any expectation that this framework would try to address these or should the user plan to script this outside of serverless?

> 674 Scenario	7:	Project	Sync
> 675 Sync remote	project	from	local
> 676 The user has already started a project (manifest) and deployed it. They have modified the
> 677 manifest by adding, removing or updating existing entities and wants to re-deploy the project.
> 678 The local addition, deletion or update of these affected entities should be reflected in the remote
> 679 OpenWhisk platform.
> 680 Sync	local	project	from	remote
> 681 The user has already deployed a project (manifest) and to a remote OpenWhisk platform. They
> 682 have modified (i.e., added, updated or deleted entities) in the remotely deployed project (perhaps
> 683 using the remote platforms UI or the command line interface (CLI). The remote addition,
> 684 deletion or update of these affected entities should be reflected in the remote OpenWhisk
> 685 platform.

**Comment by jthomas** [comment]
The Serverless Framework treats the `serverless.yml` as the "ground source of truth". Any changes to the remote resources independently will need to be replicated in that file.

Does that make sense?

**Comment by allen-servedio** [comment]
Yes, totally understand. Maybe to give a more concrete example... Suppose I initially define an action called `myAction` in `serverless.yml`. I later decide that I no longer need `myAction` and remove it from `serverless.yml`. So far as I can tell, serverless framework essentially orphans the action as it no longer knows about it. I think there are a LOT of use cases where this just makes sense... But there may be some in which they want to manage everything through the framework and so would have wanted that action to be removed.

I just wanted to verify that in cases where an action is deleted or there is a change in its name (which could look like a delete of the old one after serverless deploys the new name), that should be scripted outside of the serverless framework.

Basically wanted to make sure I was not missing something obvious.

**Comment by jthomas** [comment]
```
So far as I can tell, serverless framework essentially orphans the action as it no longer knows about it. 
```

That's correct. It won't attempt to remove resources not listed in the YAML. You could probably write a serverless plugin to check the deployed resources against the config and warn or clean up automatically.

Once we get all resources being deployed within a package, it should be possible to have a configuration setting to say "clean out resources on deployment". At the moment, we have to deploy triggers and rules in the root namespace so we don't know what is a project resource and what is something else.
